tests/phpunit/api/v3/MessageTemplateTest.php

```diff
@@ -77,102 +69,122 @@ public function testDelete() {
-      1 => [$wfName, 'String'],
+  public function testWorkflowNameToID(): void {
+    $workflowName = 'petition_sign';
+    $workflowID = CRM_Core_DAO::singleValueQuery('SELECT id FROM civicrm_option_value WHERE name = %1', [
+      1 => ['petition_sign', 'String'],
     ]);
 
     $created = $this->callAPISuccess('MessageTemplate', 'create', [
       'msg_title' => __FUNCTION__,
       'msg_subject' => __FUNCTION__,
       'msg_text' => __FUNCTION__,
       'msg_html' => __FUNCTION__,
-      'workflow_name' => $wfName,
+      'workflow_name' => $workflowName,
     ]);
-    $this->assertEquals($wfName, $created['values'][$created['id']]['workflow_name']);
-    $this->assertEquals($wfId, $created['values'][$created['id']]['workflow_id']);
+    $this->assertEquals($workflowName, $created['values'][$created['id']]['workflow_name']);
+    $this->assertEquals($workflowID, $created['values'][$created['id']]['workflow_id']);
     $get = $this->callAPISuccess('MessageTemplate', 'getsingle', ['id' => $created['id']]);
-    $this->assertEquals($wfName, $get['workflow_name']);
-    $this->assertEquals($wfId, $get['workflow_id']);
+    $this->assertEquals($workflowName, $get['workflow_name']);
+    $this->assertEquals($workflowID, $get['workflow_id']);
   }
 
-  public function testPermissionChecks() {
-    $entity = $this->createTestEntity();
+  /**
+   * Test workflow permissions.
+   *
+   * edit message templates allows editing all templates, otherwise:
+   * - edit user-driven message templates is required when workflow_name is not set.
+   * - edit system workflow message templates is required when workflow_name is set.
+   */
+  public function testPermissionChecks(): void {
+    $this->createTestEntity('MessageTemplate', [
+      'msg_title' => 'title',
+      'msg_subject' => 'subject',
+      'msg_html' => 'html',
+      'workflow_name' => 'friend',
+    ], 'workflow');
+
+    $this->createTestEntity('MessageTemplate', [
+      'msg_title' => 'title',
+      'msg_subject' => 'subject',
+      'msg_html' => 'html',
+    ], 'user');
+
     CRM_Core_Config::singleton()->userPermissionClass->permissions = ['edit user-driven message templates'];
-    // Ensure that it cannot create a system message or update a system message tempalte given current permissions.
+    // Attempting to update the workflow template should fail with only user permissions.
     $this->callAPIFailure('MessageTemplate', 'create', [
-      'id' => $entity['id'],
+      'id' => $this->ids['MessageTemplate']['workflow'],
       'msg_subject' => 'test msg permission subject',
       'check_permissions' => TRUE,
     ]);
-    $testUserEntity = $entity['values'][$entity['id']];
-    unset($testUserEntity['id']);
-    $testUserEntity['msg_subject'] = 'Test user message template';
-    unset($testUserEntity['workflow_id']);
-    unset($testUserEntity['workflow_name']);
-    $testuserEntity['check_permissions'] = TRUE;
```

Comment:
This variable should be `$testUserEntity` - but when I fixed it it started failing - which is what set me on the path of trying to make sense of this test / clean it up